File: src/main/java/org/oewntk/wndb/in/SenseToVerbTemplatesParser.java
/*
 * Copyright (c) 2021. Bernard Bou.
 */

package org.oewntk.wndb.in;

import kotlin.Pair;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Collection;

/**
 * Sense-to-verb_templates parser
 */
public class SenseToVerbTemplatesParser
{
	private final File inDir;

	/**
	 * Constructor
	 *
	 * @param inDir extra WNDB dir
	 */
	public SenseToVerbTemplatesParser(final File inDir)
	{
		this.inDir = inDir;
	}

	/**
	 * Parse verb templates per sense
	 *
	 * @return collection of sensekey-verb_templates_ids pairs
	 * @throws IOException io exception
	 */
	public Collection<Pair<String, Integer[]>> parse() throws IOException
	{
		Collection<Pair<String, Integer[]>> result = new ArrayList<>();
		parseVerbTemplates(new File(inDir, "sentidx.vrb"), result);
		return result;
	}

	/**
	 * Parse verb templates per sense
	 *
	 * @param file    file
	 * @param entries accumulator of sensekey-verb_templates_ids pairs
	 * @throws IOException io exception
	 */
	private static void parseVerbTemplates(File file, Collection<Pair<String, Integer[]>> entries) throws IOException
	{
		// iterate on lines
		try (BufferedReader reader = new BufferedReader(new InputStreamReader(new FileInputStream(file), StandardCharsets.UTF_8)))
		{
			int lineCount = 0;
			String line;
			while ((line = reader.readLine()) != null)
			{
				lineCount++;
				if (line.isEmpty() || line.charAt(0) == ' ')
				{
					continue;
				}

				try
				{
					String[] fields = line.split("[\\s,]+");
					String sensekey = fields[0];
					Integer[] templateIds = new Integer[fields.length - 1];
					for (int i = 1; i < fields.length; i++)
					{
						templateIds[i - 1] = Integer.parseInt(fields[i]);
					}
					entries.add(new Pair<>(sensekey, templateIds));
				}
				catch (final RuntimeException e)
				{
					Tracing.psErr.println("[E] verb templates at line " + lineCount + " " + e);
				}
			}
		}
	}
}

File: src/main/java/org/oewntk/wndb/in/SenseToVerbTemplatesParser.kt
/*
 * Copyright (c) 2021-2024. Bernard Bou.
 */
package org.oewntk.wndb.`in`

import org.oewntk.model.SenseKey
import java.io.*
import java.nio.charset.StandardCharsets

/**
 * Sense-to-verb_templates parser
 *
 * @property inDir extra WNDB dir
 */
class SenseToVerbTemplatesParser(
    private val inDir: File,
) {

    /**
     * Parse verb templates per sense
     *
     * @return collection of sensekey-verb_templates_ids pairs
     * @throws IOException io exception
     */
    @Throws(IOException::class)
    fun parse(): List<Pair<SenseKey, List<Int>>> {
        val result: MutableList<Pair<SenseKey, List<Int>>> = ArrayList()
        parseVerbTemplates(File(inDir, "sentidx.vrb"), result)
        return result
    }

    companion object {

        /**
         * Parse verb templates per sense
         *
         * @param file    file
         * @param entries accumulator of sensekey-verb_templates_ids pairs
         * @throws IOException io exception
         */
        @Throws(IOException::class)
        private fun parseVerbTemplates(file: File, entries: MutableList<Pair<SenseKey, List<Int>>>) {
            // iterate on lines
            BufferedReader(InputStreamReader(FileInputStream(file), StandardCharsets.UTF_8)).use { reader ->
                var lineCount = 0
                reader.useLines { lines ->
                    lines.forEach { line ->
                        lineCount++
                        if (line.isNotEmpty() || line[0] != ' ') {
                            try {
                                val fields = line.split("[\\s,]+".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                                val senseKey = SenseKey(fields[0])
                                val templateIds = fields.drop(1).map(String::toInt).toList()
                                entries.add(Pair(senseKey, templateIds))
                            } catch (e: RuntimeException) {
                                Tracing.psErr.println("[E] verb templates at line $lineCount $e")
                            }
                        }
                    }
                }
            }
        }
    }
}
